src/main/scala/analysis/SummaryGenerator.scala

```diff
@@ -0,0 +1,186 @@
+    with SimpleWorklistFixpointSolver[CFGPosition, Set[Taintable], PowersetLattice[Taintable]]
+
+/**
+ * Generates summaries (requires and ensures clauses) for procedures that necessarily must hold (assuming a correct implementation).
+ * This helps because the verifier cannot make any assumptions about procedures with no summaries.
+ */
+class SummaryGenerator(
+    program: Program,
+    specGlobals: Set[SpecGlobal],
+    globals: Map[BigInt, String],
+    constProp: Map[CFGPosition, Map[Variable, FlatElement[BitVecLiteral]]],
+    varDepsSummaries: Map[Procedure, Map[Taintable, Set[Taintable]]],
+) {
+  val rnaResults = RNATaintableSolver(program, globals, constProp).analyze()
+
+  // TODO should the stack/link/frame pointers propagate taint?
+  val variables: Set[analysis.Taintable] = 0.to(28).map { n =>
+    Register(s"R$n", 64)
+  }.toSet ++ specGlobals.map { g =>
+    analysis.GlobalVariable(dsl.mem, BitVecLiteral(g.address, 64), g.size, g.name)
+  }
+
+  private def toGamma(variable: Taintable): Option[BExpr] = {
+    variable match {
+      case variable: Register => Some(variable.toGamma)
+      case variable: LocalVar => None
+      case variable: GlobalVariable => Some(variable.toGamma)
+      //case variable: LocalStackVariable => None
+      case variable: UnknownMemory => Some(FalseBLiteral)
+    }
+  }
+
+  private def toExprGamma(variable: Taintable): Option[BExpr] = {
+    variable match {
+      case variable: GlobalVariable => Some(BinaryBExpr(BoolOR, variable.toGamma, variable.L))
+      case _ => toGamma(variable)
+    }
+  }
+
+  /**
+   * Get a map of variables to variables which have tainted it in the procedure.
+   */
+  private def getTainters(procedure: Procedure, variables: Set[Taintable]): Map[Taintable, Set[Taintable]] = {
+    varDepsSummaries.getOrElse(procedure, Map())
+  }
+
+  /**
+   * Generate requires clauses for a procedure. Currently this does nothing.
+   */
+  def generateRequires(procedure: Procedure): List[BExpr] = List()
+
+  /**
+   * Generate ensures clauses for a procedure. Currently, all generated ensures clauses are of the form (for example)
+   *   ensures Gamma_R2 || (Gamma_R2 == old(Gamma_y)) || (Gamma_R2 == old(Gamma_R1)) || (Gamma_R2 == old(Gamma_R2));
+   * whenever this can be done soundly.
+   */
+  def generateEnsures(procedure: Procedure): List[BExpr] = {
+    if procedure.blocks.isEmpty then return List()
+
+    // We only need to make ensures clauses about the gammas of modified variables.
+    val relevantVars = variables.filter { v =>
+      v match {
+        case v: Global => procedure.modifies.contains(v)
+        case _ => true
+      }
+    }
+
+    // Use rnaResults to find stack function arguments
+    val tainters = relevantVars.map {
+      v => (v, Set())
+    }.toMap ++ getTainters(procedure, variables ++ rnaResults(procedure.begin) + UnknownMemory()).filter { (variable, taints) =>
+      relevantVars.contains(variable)
+    }
```

Comment:
A comment explaining what this represents and how the sets of tainters for each variable have been derived would help.